contracts/Market.sol

```diff
@@ -0,0 +1,857 @@
+  }
+
+  /**
+   * @dev The distinct removal token ids listed in the market that are reserved.
+   */
+  function reservedSupply() external view returns (uint256[] memory) {
+    return _reservedSupply.values();
+  }
+
+  /**
+   * @dev Reverts if market is out of stock or if available stock is being reserved for priority buyers
+   * and buyer is not priority.
+   *
+   * @param purchaseAmount The number of carbon removals being purchased.
+   */
+  function _checkSupply(uint256 purchaseAmount) private view {
+    // TODO: BUG: when using swap from single supplier, this function should check against the suppliers active balance,
+    //not the markets total active balance!
+    // TODO: BUG: this function doesn't consider reserved removals!
+    uint256 activeSupply = _removal.cumulativeBalanceOf(address(this));
+    if (activeSupply == 0) {
+      revert OutOfStock();
+    }
+    if (activeSupply <= _priorityRestrictedThreshold) {
+      if (!hasRole(ALLOWLIST_ROLE, _msgSender())) {
+        revert LowSupplyAllowlistRequired();
+      }
+    }
+    if (purchaseAmount > activeSupply) {
+      revert InsufficientSupply(); // todo test the logic for these checks is correct for all states
+    }
+  }
+
+  /**
+   * @dev Determines the removal ids, amounts, and suppliers to fill the given purchase quantity in
+   * a round-robin order.
+   *
+   * @param certificateAmount The number of carbon removals to purchase.
+   * @return numberOfRemovalForOrder The number of distinct removal token ids used to fulfill this order.
+   * @return ids An array of the removal token ids being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplyRoundRobin(uint256 certificateAmount)
+    private
+    returns (
+      uint256,
+      uint256[] memory,
+      uint256[] memory,
+      address[] memory
+    )
+  {
+    uint256 remainingAmountToFill = certificateAmount;
+    uint256 numberOfActiveRemovalsInMarket = _removal
+      .numberOfTokensOwnedByAddress(
+        address(this) // todo are we making calls like this throughout the contract flow? if so can we pass values down?
+      );
+    uint256[] memory ids = new uint256[](numberOfActiveRemovalsInMarket);
+    uint256[] memory amounts = new uint256[](numberOfActiveRemovalsInMarket);
+    address[] memory suppliers = new address[](numberOfActiveRemovalsInMarket);
+    uint256 numberOfRemovalsForOrder = 0;
+    // TODO (Gas Optimization): Declare variables outside of loop
+    for (
+      uint256 i = 0;
+      i < numberOfActiveRemovalsInMarket;
+      i++ // todo ++i consistency
+    ) {
+      uint256 removalId = _activeSupply[_currentSupplierAddress]
+        .getNextRemovalForSale();
+      uint256 removalAmount = _removal.balanceOf(address(this), removalId); // todo batch
```

Comment:
We might also considering moving most everything from here down info a helper with a signature along the lines of:

`sellQuantityFromRemoval(uint256 quantity, RemovalID removalId)`